service_discovery/core/src/registry_impl.rs

```diff
@@ -0,0 +1,457 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+// SPDX-License-Identifier: MIT
+
+//! Module containing gRPC service implementation based on [`proto_servicediscovery::chariott_registry::v1`].
+//!
+//! Provides a gRPC endpoint for external services to interact with to register and discover
+//! services. Note: Identifiers in all Registry operations are case-sensitive.
+//!
+use parking_lot::{RwLock, RwLockReadGuard, RwLockWriteGuard};
+use proto_servicediscovery::chariott_registry::v1::registry_server::Registry;
+use proto_servicediscovery::chariott_registry::v1::{
+    DiscoverByNamespaceRequest, DiscoverByNamespaceResponse, DiscoverServiceRequest,
+    DiscoverServiceResponse, InspectRequest, InspectResponse, RegisterRequest, RegisterResponse,
+    RegistrationStatus, ServiceMetadata, UnregisterRequest, UnregisterResponse,
+};
+use std::collections::HashMap;
+use std::sync::Arc;
+use std::vec::Vec;
+use tonic::{Request, Response, Status};
+use tracing::{info, warn};
+
+/// Base structure for the registry gRPC service
+#[derive(Clone, Debug)]
+pub struct RegistryImpl {
+    pub registry_map: Arc<RwLock<HashMap<ServiceIdentifiers, ServiceMetadata>>>,
```

Comment:
do we need to expose this externally? we can have specific methods on the registry for add/remove/get that, internally, access this specific structure